### ISSUE TITLE
Implement Debug Adapter Protocol

### ISSUE THREAD

**Issue by p-bakker** [mozilla/rhino#973]
In order to be able to use Rhino in modern develoment setups, I propose to implement the [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/), which is how [VS Code](https://code.visualstudio.com/api/extension-guides/debugger-extension) and [Wild Web Developer (Eclipse)](https://github.com/eclipse/wildwebdeveloper) offer an extendible debug targets

I wonder if anyone uses the Swing-based debugger that in part of Rhino and I know there's a implementation based on the [dbgp](https://xdebug.org/docs/dbgp) protocol in [DTLK](https://github.com/eclipse/dltk.javascript/tree/master/plugins/org.eclipse.dltk.javascript.rhino.dbgp), which is used by some, but all platforms nowadays seem to move towards DAP

**Comment by p-bakker** [comment]
Within Java land, there's [lsp4e](https://projects.eclipse.org/projects/technology.lsp4e), to aid easier implementation of both DAP and Language Servers

**Comment by p-bakker** [comment]
Curious to know if anyone uses any form of a debugger for debugging just JavaScript running in Rhino

**Comment by c3-mgruchacz** [comment]
Hi @p-bakker, wondering if there's any update on implementing Debug Adapter Protocol for Rhino?

Thanks!

**Comment by p-bakker** [comment]
While I wish I had time to work on this, I haven't. I created this case as part of establishing idea's for the Rhino roadmap.

If you're interested in contributing, I'm happy to share my ideas

**Comment by tonygermano** [comment]
> I wonder if anyone uses the Swing-based debugger that in part of Rhino 
I do occasionally, but I wouldn't say I'm a heavy user. I have it configured as a build task in vscode so I can launch it from there.

**Comment by c3-mgruchacz** [comment]
Yes, I'm interested, I would like to hear your ideas. I think the lsp4j project that you reference in post 2 is a good starting point, essentially implement the IDebugProtocolServer interface, mapping each method to the corresponding action in the rhino debugger class Dim.

**Comment by p-bakker** [comment]
Great to hear!

Yeah, I think using lsp4j as a basis, so you can focus on just the wiring things up into Rhino would be the best approach.

As for the wiring up into Rhino, bottom-line, you'll need an implementation of org.mozilla.javascript.debug.Debugger that can be registered on a Context using .setDebugger(...)

You could start implementing org.mozilla.javascript.debug.Debugger from scratch or take an existing implementation that already has implemented a lot of features like breakpoints etc., as the Debugger interface is nothing more that a hook into the internal Rhino JavaScript callstack.

As for the existing implementations: the current Swing-based debugger in the tools section that comes with Rhino would be one starting point, but that debugger isn't that feature rich. 

As mentioned above, I know of at least one other implementation, namely the one based on the DBGP protocol in the DTLK project.

Personally, I regularly use a (custom) version of the debugger in DLTK and close to being as feature rich as the Java Debugger in Eclipse (imho the only notable omission being Drop to Frame). The source of this custom version is also available: https://github.com/Servoy/servoy-js-engine/tree/master/org.eclipse.dltk.javascript.rhino/src (the Rhino side) and https://github.com/Servoy/org.eclipse.dltk.javascript/tree/servoy/plugins/org.eclipse.dltk.javascript.debug (server/debugger side)

If I would attempt to implement DAP, I think I'd start looking to see if I could reuse large chunks of the DLTKbased implementation.

When doing so, I'd first look at whether it would be ok to copy & modify the DLTK code: I think its EPL based, but it's not clearly mentioned on the repo and even if it's EPL-based, I'm not sure what that implies with regards to copying that code and using it as a basis for something else under a different license

**Comment by p-bakker** [comment]
Just as reference:

- Another Rhino Debugger implementation: https://git.eclipse.org/c/jsdt/webtools.jsdt.git/tree/bundles/org.eclipse.wst.jsdt.debug.rhino.debugger/src/org/eclipse/wst/jsdt/debug
- And an extension of the Swing-base Debugger: https://github.com/nextgenhealthcare/connect/tree/development/server/src/org/mozilla/javascript

**Comment by p-bakker** [comment]
And some more pointers: based on [this discussion](https://groups.google.com/g/mozilla.dev.tech.js-engine.rhino/c/pNJlUd_2ueg/m/GBc2std9czEJ), at least 2  remote debugger implementations were being worked on at one point in time, one of them by @szegedi, which can still be found at https://sourceforge.net/p/dynalang/code/HEAD/tree/trunk/debug-rhino/

**Comment by p-bakker** [comment]
another Rhino Debug impl., Eclipse Ease this time: https://gitlab.eclipse.org/eclipse/ease/ease/-/tree/main/plugins/org.eclipse.ease.lang.javascript.rhino/src/org/eclipse/ease/lang/javascript/rhino/debugger?ref_type=heads

And seems to maybe already do some DAP-related stuff, as [this commit](https://gitlab.eclipse.org/eclipse/ease/ease/-/commit/b75d2841deb70c4ab5ae7c15c8c23953ee547fb1) introduces support for WildWebdeveloper breakpoints, which, from a quick glance at the changes, seem to be LSP4J breakpoints

**Comment by tonygermano** [comment]
Since my original comment two years ago, https://github.com/nextgenhealthcare/connect now integrates the swing debugger into their product (but it doesn't allow remote debugging, which limits its use as it typically runs on headless systems.)

**Comment by davidpcaldwell** [comment]
A few quick things to mention:

* I also have embedded the Rhino debugger. This is a pretty thin wrapper, but it's something folks could look at. https://github.com/davidpcaldwell/slime/blob/fb50a4f242c38e40d654f340e28c3045284178ae/loader/jrunscript/rhino/java/inonit/script/rhino/Debugger.java#L266
* Possibly more interesting is this "debugger" which uses the Rhino debugger APIs to implement a profiler. https://github.com/davidpcaldwell/slime/blob/fb50a4f242c38e40d654f340e28c3045284178ae/loader/jrunscript/rhino/java/inonit/script/rhino/Debugger.java#L52
* Finally, I have used https://github.com/provegard/ncdbg on Nashorn (since we were talking about Java implementations of the protocol). It never got to the point where it was super-stable for me, but it definitely had a lot of stuff right, as it mostly worked at one time. Not clear to me whether the protocol evolved out from under it or the JVM did, but it regressed over time.
